### PR TITLE
Added org.ardour.Ardour

### PR DESCRIPTION
**Pull request by hfiguiere**

### CODE REVIEW

**Comment by nedrichards** [comment]
bot, build org.ardour.Ardour5
testbot, build org.ardour.Ardour5

**Comment by alexlarsson** [comment]
I killed this build temporary because we had some buildbot issues

File: org.ardour.Ardour5.json [hunk 1]
```diff
@@ -0,0 +1,452 @@
+{
+  "app-id": "org.ardour.Ardour5",
```
Comment (eszlari) [inline org.ardour.Ardour5.json]: I think upstream adds the version number for parallel installs and we should drop it.

File: org.ardour.Ardour5.json [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+  "runtime": "org.gnome.Platform",
```
Comment (eszlari) [inline org.ardour.Ardour5.json]: I was working on a flatpak too and built it against freedesktop runtime successfully.

File: org.ardour.Ardour5.json [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+    "--socket=pulseaudio",
+    /* MIDI */
+    "--device=all",
+    /* Allow loading, saving files from anywhere (portals don’t work yet) */
+    "--filesystem=host",
+    /* Allow other instances to see lockfiles */
+    "--env=TMPDIR=/var/tmp"
+  ],
+  "modules": [
+    "gtk2.json",
```
Comment (eszlari) [inline org.ardour.Ardour5.json]: https://github.com/flathub/shared-modules/blob/master/gtk2/gtk2.json

File: org.ardour.Ardour5.json [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+            {
+              "type": "archive",
+              "url": "http://www.ladspa.org/download/ladspa_sdk_1.15.tgz",
+              "sha256": "4229959b09d20c88c8c86f4aa76427843011705df22d9c28b38359fd1829fded"
+            }
+          ]
+        },
+        {
+          / *Double precision build */
+          "name": "fftw3",
```
Comment (eszlari) [inline org.ardour.Ardour5.json]: This is already in the runtime, you only need the float variant.

**Comment by eszlari** [comment]
@hfiguiere Have you tested the jack support? Are you able to connect to the jack server running on the host? I couldn't get this to work yet.

**Comment by hfiguiere** [comment]
TBH this one already has HiDPI problems. It needs more testing.

File: org.ardour.Ardour5.json (hunk 1 above)
Comment (hfiguiere) [reply org.ardour.Ardour5.json]: done

File: org.ardour.Ardour5.json (hunk 4 above)
Comment (hfiguiere) [reply org.ardour.Ardour5.json]: Removed. All good.

**Comment by nedrichards** [comment]
bot, build org.ardour.Ardour5
testbot, build org.ardour.Ardour5

**Comment by hfiguiere** [comment]
manifest JSON file was renamed following the change of app-id. (title changed too)

**Comment by nedrichards** [comment]
bot, build org.ardour.Ardour
testbot, build org.ardour.Ardour

**Comment by eszlari** [comment]
> TBH this one already has HiDPI problems. It needs more testing.

This is expacted for a gtk2 based app. And it seems, they don't plan to port to gtk3 (soon):

https://discourse.ardour.org/t/a-new-modern-look-and-feel-for-ardour/87728/24
> Paul Davis Oct '15
> corebots: not happening. we’re moving away step by step from any desktop toolkit toward doing things the way we need them. GTK3 doesn’t offer us anything useful and will cause a world of pain because of plugin GUIs in GTK2.

**Comment by eszlari** [comment]
@hfiguiere Bad news: Since https://github.com/flatpak/flatpak/commit/89e2b6679cd4a59c15b225c0c6c0e1293ae83f7d flatpak doesn't support JACK anymore!

https://github.com/flatpak/flatpak/issues/1509

**Comment by hfiguiere** [comment]
I know about jack. That's gonna be a blocker for a few of the package submissions I have in the pipe.

File: org.ardour.Ardour.json [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+      ],
+      "post-install": [
+        "mkdir -p /app/share/applications/",
+        "desktop-file-install --dir /app/share/applications --set-key=Name --set-value=\"Ardour 5\" --set-generic-name=\"Digital Audio Workstation\" --set-key=X-GNOME-FullName --set-value=\"Ardour v5 (Digital Audio Workstation)\" --set-comment=\"Record, mix and master audio\" --remove-category=AudioEditing --add-category=X-AudioEditing build/gtk2_ardour/ardour5.desktop",
+        "for s in 16 22 32 48 256 512; do mkdir -p /app/share/icons/hicolor/${s}x${s}/apps ; cp /app/share/ardour5/resources/Ardour-icon_${s}px.png /app/share/icons/hicolor/${s}x${s}/apps/${FLATPAK_ID}.png ; done"
+      ],
+      "sources": [
+        {
+          "type": "archive",
+          "url": "https://community.ardour.org/srctar",
```
[the remark this answers is not in the document]
Comment (hfiguiere) [reply org.ardour.Ardour.json]: I am not sure the archive is the same content. I will check.

I have asked upstream about where to find the tarball archive and I was told to use tags in git... I didn't seem to have a convincing argument as to why tarballs of older releases are needed.

File: org.ardour.Ardour.json [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+      "build-commands": [
+        "./waf configure --prefix=/app --no-phone-home --freedesktop --with-backends=jack,alsa,dummy",
+        "./waf build -j $FLATPAK_BUILDER_N_JOBS",
+        "./waf i18n -j $FLATPAK_BUILDER_N_JOBS",
+        "./waf install"
+      ],
+      "post-install": [
+        "mkdir -p /app/share/applications/",
+        "desktop-file-install --dir /app/share/applications --set-key=Name --set-value=\"Ardour 5\" --set-generic-name=\"Digital Audio Workstation\" --set-key=X-GNOME-FullName --set-value=\"Ardour v5 (Digital Audio Workstation)\" --set-comment=\"Record, mix and master audio\" --remove-category=AudioEditing --add-category=X-AudioEditing build/gtk2_ardour/ardour5.desktop",
+        "for s in 16 22 32 48 256 512; do mkdir -p /app/share/icons/hicolor/${s}x${s}/apps ; cp /app/share/ardour5/resources/Ardour-icon_${s}px.png /app/share/icons/hicolor/${s}x${s}/apps/${FLATPAK_ID}.png ; done"
```
[the remark this answers is not in the document]
Comment (hfiguiere) [reply org.ardour.Ardour.json]: ok

File: org.ardour.Ardour.json [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+      "name": "ardour",
+      "buildsystem": "simple",
+      "build-commands": [
+        "./waf configure --prefix=/app --no-phone-home --freedesktop --with-backends=jack,alsa,dummy",
+        "./waf build -j $FLATPAK_BUILDER_N_JOBS",
+        "./waf i18n -j $FLATPAK_BUILDER_N_JOBS",
+        "./waf install"
+      ],
+      "post-install": [
+        "mkdir -p /app/share/applications/",
```
[the remark this answers is not in the document]
Comment (hfiguiere) [reply org.ardour.Ardour.json]: ok

File: org.ardour.Ardour.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (hfiguiere) [reply org.ardour.Ardour.json]: This doesn't work. Basically the archive is empty. It contain the following 

````
Github release tarballs are not supported by Ardour
===================================================

This information applies to bundles automatically generated by
https://github.com/Ardour/ardour/releases

Ardour's program version is maintained using git-tags. Compiling Ardour depends
on `git describe` (a full git clone is needed for this) or alternatively a
version-file in the release-dist. Github cannot provide the latter and also
does not allow to inject a version-file during export archiving.

For further information see http://tracker.ardour.org/view.php?id=7328
````

[the remark this answers is not in the document]
Comment (hfiguiere) [reply org.ardour.Ardour.json]: However I found a proper link to the tarball. Will use that.

[the remark this answers is not in the document]
Comment (scx) [reply org.ardour.Ardour.json]: @hfiguiere 

> This doesn't work. Basically the archive is empty.

You can always use the Git tag.
```
        {
          "type": "git",
          "url": "https://github.com/Ardour/ardour.git",
          "tag": "5.12",
          "commit": "ae0dcdc0c5d13483271065c360e378202d20170a"
        },
```

**Comment by scx** [comment]
@nedrichards 
It is the `org.ardour.Ardour` now, not `org.ardour.Ardour5`.

**Comment by scx** [comment]
testbot, build org.ardour.Ardour

**Comment by scx** [comment]
Builds fine, except for AppData/desktop/icon.
```
stat: cannot stat ‘builddir/*/share/app-info/xmls/org.ardour.Ardour.xml.gz’: No such file or directory
```

File: org.ardour.Ardour.json [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+        "desktop-file-install --dir /app/share/applications --set-key=Name --set-value=\"Ardour 5\" --set-generic-name=\"Digital Audio Workstation\" --set-key=X-GNOME-FullName --set-value=\"Ardour v5 (Digital Audio Workstation)\" --set-comment=\"Record, mix and master audio\" --remove-category=AudioEditing --add-category=X-AudioEditing build/gtk2_ardour/ardour5.desktop",
```
[the remark this answers is not in the document]
Comment (hfiguiere) [reply org.ardour.Ardour.json]: done

**Comment by scx** [comment]
testbot, build org.ardour.Ardour

**Comment by scx** [comment]
@hfiguiere 
Still the same error.
For some reason, the AppData file has not been installed.

As you can see, there is a wscript that translates the desktop file and the AppData file.
https://github.com/Ardour/ardour/blob/ae0dcdc0c5d13483271065c360e378202d20170a/gtk2_ardour/wscript#L835-L856

However, only the first one is being installed.
https://github.com/Ardour/ardour/blob/ae0dcdc0c5d13483271065c360e378202d20170a/tools/linux_packaging/stage2.run.in#L520-L531
https://github.com/Ardour/ardour/blob/ae0dcdc0c5d13483271065c360e378202d20170a/tools/linux_packaging/stage2.run.in#L593-L620

So, it seems that you will have to install it manually in the `post-install` section.
```
        "install -Dm644 build/gtk2_ardour/ardour5.appdata.xml /app/share/appdata/${FLATPAK_ID}.appdata.xml",
```

Moreover, you still have to change value of the `Icon` filed in the desktop file as I mentioned earlier.
https://github.com/flathub/flathub/pull/831#discussion_r254093159

File: org.ardour.Ardour.json [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+      "name": "ardour",
+      "buildsystem": "simple",
+      "build-commands": [
+        "./waf configure --prefix=/app --no-phone-home --freedesktop --with-backends=jack,alsa,dummy",
+        "./waf build -j $FLATPAK_BUILDER_N_JOBS",
+        "./waf i18n -j $FLATPAK_BUILDER_N_JOBS",
+        "./waf install"
+      ],
+      "post-install": [
+        "desktop-file-edit --set-key=Name --set-value=\"Ardour 5\" --set-generic-name=\"Digital Audio Workstation\" --set-key=X-GNOME-FullName --set-value=\"Ardour v5 (Digital Audio Workstation)\" --set-comment=\"Record, mix and master audio\" --remove-category=AudioEditing --add-category=X-AudioEditing build/gtk2_ardour/ardour5.desktop",
```
Comment (scx) [inline org.ardour.Ardour.json]: Why you are doing it here? I mean, you have already provided the whole `ardour5.desktop` file, so this is questionable.

In my opinion, you should either:
1. Just provide modified `ardour.desktop.in` file in sources (just remember about `dest`: `gtk2_ardour`) and don't use the `desktop-file-edit` command here.
2. Don't provide `ardour.desktop.in` nor `ardour5.desktop` file, and just edit auto-generated `build/gtk2_ardour/ardour5.desktop` file in post-install section using `desktop-file-edit` command.

File: ardour5.desktop [hunk 10]
```diff
@@ -0,0 +1,12 @@
+[Desktop Entry]
```
Comment (scx) [inline ardour5.desktop]: Why is it here, if you later use the `desktop-file-edit` command?

File: org.ardour.Ardour.json (hunk 9 above)
Comment (scx) [reply org.ardour.Ardour.json]: Upon reflection, I would just add `-set-icon=${FLATPAK_ID}` here:
```
        "desktop-file-edit --set-key=Name --set-value=\"Ardour 5\" --set-generic-name=\"Digital Audio Workstation\" --set-key=X-GNOME-FullName --set-value=\"Ardour v5 (Digital Audio Workstation)\" --set-comment=\"Record, mix and master audio\" --remove-category=AudioEditing --add-category=X-AudioEditing --set-icon=${FLATPAK_ID} build/gtk2_ardour/ardour5.desktop",
```

Comment (scx) [reply org.ardour.Ardour.json]: Of course you still have to install the AppData file manually, as I mentioned earlier.
https://github.com/flathub/flathub/pull/831#issuecomment-461332534

**Comment by nedrichards** [comment]
@hfiguiere would you say this is blocked by the lack of JACK support in flatpak? Flathub doesn't want to ship half working software, it just generates frustration and hurts everyone else.

**Comment by scx** [comment]
@nedrichards 

> @hfiguiere would you say this is blocked by the lack of JACK support in flatpak? Flathub doesn't want to ship half working software, it just generates frustration and hurts everyone else.

Well, we could say that about every IDE here (maybe except GNOME Builder, which is designed for flatpak).
If he somehow managed to make it work with the `pipewire-jack`, I think that it should be accepted. At least I don't see any contraindications.
https://github.com/PipeWire/pipewire-jack

File: ardour5.desktop (hunk 10 above)
Comment (hfiguiere) [reply ardour5.desktop]: no idea why I added it to git. Will remove

**Comment by hfiguiere** [comment]
Jack is an issue, indeed. And yes I'm ok to wait on that since it's required.

File: org.ardour.Ardour.json (hunk 9 above)
Comment (hfiguiere) [reply org.ardour.Ardour.json]: done

File: ardour5.desktop (hunk 10 above)
Comment (hfiguiere) [reply ardour5.desktop]: removed.

**Comment by scx** [comment]
testbot, build org.ardour.Ardour

**Comment by hfiguiere** [comment]
I know what's going on.

**Comment by scx** [comment]
testbot, build org.ardour.Ardour

**Comment by hfiguiere** [comment]
I re-tested this. I'm not sure exactly why I reached the previous conclusion, but Ardour work without Jack, using ALSA.

**Comment by nedrichards** [comment]
@hfiguiere oh! ok ;-) what does upstream think about this? are they interested in getting involved? is there a thread or bug you can point to?

**Comment by hfiguiere** [comment]
Didn't talk to upstream. Their download page for builds request payment.

**Comment by hfiguiere** [comment]
The only question I asked upstream is if there is a way to be able to use Ardour on HiDPI, since by default even the start dialog is unreadable.

**Comment by eszlari** [comment]
Pinging upstream: @pauldavisthefirst

**Comment by hfiguiere** [comment]
I might have a hackish solution for the screen resolution. Will involve guess work and messing with the config file on startup. Can't validate until next week though.

**Comment by eszlari** [comment]
Maybe this could be added as beta first like Lutris (https://github.com/flathub/net.lutris.Lutris)?

It could be helpful to @wtay as a test case for PipeWire's JACK support (https://blogs.gnome.org/uraeus/2019/09/23/fedora-workstation-31-whats-new/).

**Comment by hfiguiere** [comment]
I need to update this to use the newer SDK to see with pipewire

**Comment by hfiguiere** [comment]
How does one use pipewire in that case?

**Comment by pauldavisthefirst** [comment]
I've muted this thread, because I don't have the bandwidth to deal with it, and it's not something we're really that interested in (3rd party packaging of Ardour in general, not just Flatpak). 

Ardour runs on HiDPI displays on macOS and I don't know why it would be a problem on Linux, though you may have to adjust GUI scaling in preferences. 

If you need to come talk to us, IRC #ardour on freenode is the best avenue. rgareus is another lead dev.

**Comment by hfiguiere** [comment]
Re: HiDPI here is a summary

1. I can't read any bits of the UI by default, it is far too small.
2. Therefore it is hard to go to the preference and guess what to change.

When I asked on #ardour what config file to edit to change that, all I heard was crickets. I could indeed read the the source code and try to figure it out, but I didn't have the time for that.

BTW that's not what's blocking here anyway, so it becomes off topic.

Also @eszlari pinging people like that is RUDE.

**Comment by hfiguiere** [comment]
bot, build

**Comment by hfiguiere** [comment]
bot, build org.ardour.Ardour

**Comment by hfiguiere** [comment]
This latest change address the issue of HiDPI screens as it detect the resolution.

**Comment by hfiguiere** [comment]
bot, build org.ardour.Ardour

**Comment by hfiguiere** [comment]
So I have made the decision to build without jack since jack doesn't work in flatpak, and Ardour can do without. 

Now the problem is how to bring in plugins.

**Comment by LeandroStanger** [comment]
bot, build org.ardour.Ardour

**Comment by LeandroStanger** [comment]
bot, build org.ardour.Ardour

**Comment by AsciiWolf** [comment]
Any progress?

**Comment by hfiguiere** [comment]
flatpak/flatpak#3364 should allow us to go forward and work properly with Jack.

**Comment by eszlari** [comment]
https://news.ycombinator.com/item?id=22393786

> To be fair, as the original author of JACK and the lead author of one the main DAWs on Linux, we no longer encourage the use of JACK with Ardour.
> 
> JACK is an exceptional powerful tool (if I do say so myself) but it is overkill for the majority (maybe even the vast majority) of users. We try to encourage most Ardour users to use its builtin ALSA audio/MIDI I/O support rather than JACK these days.

**Comment by nedrichards** [comment]
So this could set "--require-version=1.6.1", with the --device=shm permission and be OK? 1.6.x looks to be about 15% of flathub traffic currently.

**Comment by hfiguiere** [comment]
I can't even build with these flags as I only 1.4.4

**Comment by nedrichards** [comment]
Yeah, that was more of a "I believe this is theoretically possible" but you're likely to want to wait a while!

**Comment by hfiguiere** [comment]
Update on this Pull Request:

There is an extension point to build LV2 plugins to be usable by Ardour. I will submit one soon that has been tested.

I am pondering if we couldn't have a way to have all LV2 plugins usable by any flatpak that supports LV2. I'm not sure of the approach to take. Muse3 could support them too.

**Comment by hfiguiere** [comment]
Just this that greatly increase the usability of Ardour in Flatpak form.

**Comment by hfiguiere** [comment]
Sample plugin part of pr #1459

**Comment by hfiguiere** [comment]
bot, build org.ardour.Ardour

**Comment by eszlari** [comment]
> I can't even build with these flags as I only 1.4.4

You could leave out `--require-version=1.6.1` and just add `--device=shm` (unsupported flags will be ignored), so that JACK is optional (as quoted above: Ardour's creator doesn't see it as essential).

> I am pondering if we couldn't have a way to have all LV2 plugins usable by any flatpak that supports LV2. I'm not sure of the approach to take.

`org.freedesktop.Platform.LV2` ?

**Comment by hfiguiere** [comment]
> > I can't even build with these flags as I only 1.4.4
> 
> You could leave out `--require-version=1.6.1` and just add `--device=shm` (unsupported flags will be ignored), so that JACK is optional (as quoted above: Ardour's creator doesn't see it as essential).
 
Things have moved since (the comment is from Feb 28). If you look at the JSON, there is no such flag. But `--device=shm` is there. I add Jack as an option and we can let the user decide. I think it's a decent compromise. I realize in the end I haven't tested it, I used ALSA.

> > I am pondering if we couldn't have a way to have all LV2 plugins usable by any flatpak that supports LV2. I'm not sure of the approach to take.
> 
> `org.freedesktop.Platform.LV2` ?

Does that exist? Or are you suggesting we create it? That wouldn't be a bad idea.

**Comment by eszlari** [comment]
> Does that exist? Or are you suggesting we create it?

We would have to create it. But should be rather `org.kde.Platform.LV2`, because while most plugins seems to use gtk/gtkmm, they are some that use qt5.

**Comment by eszlari** [comment]
Just discovered this old thread: http://lists.lv2plug.in/pipermail/devel-lv2plug.in/2016-March/001593.html

The situation with LV2 and GUI toolkits is really a mess.

**Comment by hfiguiere** [comment]
> Just discovered this old thread: http://lists.lv2plug.in/pipermail/devel-lv2plug.in/2016-March/001593.html
> 
> The situation with LV2 and GUI toolkits is really a mess.

And then when I use anything that is not Gtk3 or Qt5 on my 3 year old laptop with an HiDPI screen, I need to pull out the magnifier to read the text.

**Comment by hfiguiere** [comment]
I will update this PR to use a base app to build and provide extension support. Stay tuned.

**Comment by hfiguiere** [comment]
For some reason I have sound issues. But then it is not the only app that has issues.

**Comment by hfiguiere** [comment]
> For some reason I have sound issues. But then it is not the only app that has issues.

1. I think my system has issues there, which isn't encouraging
2. I updated last night with Ardour 6 (rc1) and it actually works, so I'm willing to wait for the final release. I'll update the PR tonight with the rc1.

**Comment by hfiguiere** [comment]
With this last patch, I enabled device reservation, Ardour can be run without `pasuspender` and I get sound in an out using ALSA.

**Comment by hfiguiere** [comment]
In light of this, I stripped jack out of this build. We can reconsider when `pw-jack` is working

**Comment by LeandroStanger** [comment]
bot, build org.ardour.Ardour

**Comment by Bleuzen** [comment]
@hfiguiere Ardour 6.0 is tagged now (https://github.com/Ardour/ardour/releases/tag/6.0).
Please update this flatpak.
Hope to see it soon on Flathub :)

**Comment by hfiguiere** [comment]
Apparently I don't get notified of new tag/release through github.... *sigh*

Will definitely update.

**Comment by eszlari** [comment]
> Apparently I don't get notified of new tag/release through github.... sigh

Unfortunately, a tag is not enough to get notified, they need to make a Github release.

**Comment by hfiguiere** [comment]
I was expecting too much from Github. Any way, this is updated to 6.0

**Comment by hfiguiere** [comment]
bot, build org.ardour.Ardour

**Comment by hfiguiere** [comment]
Package is working. This is IMHO ready.

Then we'll need plugins, there is a bunch waiting as pull request.

**Comment by hfiguiere** [comment]
Turns out I need to add `--optimize --debug-symbols` to the configure: optimized but with debug symbols.

**Comment by hfiguiere** [comment]
> > Apparently I don't get notified of new tag/release through github.... sigh
> 
> Unfortunately, a tag is not enough to get notified, they need to make a Github release.

As I understand they are waiting on the builds to announce the release.

**Comment by hfiguiere** [comment]
bot, build org.ardour.Ardour

**Comment by hfiguiere** [comment]
The aarch64 build did timeout: https://flathub.org/builds/#/builders/51/builds/1478 :-/

**Comment by hfiguiere** [comment]
bot, build org.ardour.Ardour

**Comment by hfiguiere** [comment]
@nedrichards this is no longer work in progress. This is now working.

tl;dr: it is Ardour 6 now, it uses ASLA (Pulse audio option) and not jack, and it support the Linux Audio plugin as packaged in Flathub.

File: gtkmm2.json [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        "/lib/gdkmm-2.4",
+        "*.la"
+    ],
+    "config-opts" : [
+        "--disable-documentation"
+    ],
+    "sources": [
+        {
+            "type": "archive",
+            "url": "http://ftp.gnome.org/pub/GNOME/sources/gtkmm/2.24/gtkmm-2.24.5.tar.xz",
```
Comment (bilelmoussaoui) [inline gtkmm2.json:106]: Please use https://download.gnome.org/ instead

**Comment by bilelmoussaoui** [comment]
Have you reached to upstream to see if they are open to (co-)maintain Ardour here on Flathub?

**Comment by hfiguiere** [comment]
> Have you reached to upstream to see if they are open to (co-)maintain Ardour here on Flathub?

TBF I'm afraid to ask:
https://github.com/flathub/flathub/pull/831#issuecomment-472080698

Also upstream was poked (not by me) and I commented here, so they are aware:
https://github.com/flathub/flathub/pull/831#issuecomment-536245219

Now that Ardour 6 is it, it will be back to a bi-monthly release schedule and since they tend to be self-contained, I'm confident it will be manageable.

**Comment by hfiguiere** [comment]
I had a short conversation about it on IRC and it's a negative.

**Comment by hfiguiere** [comment]
bot, build org.ardour.Ardour

File: gtkmm2.json (hunk 11 above)
Comment (hfiguiere) [reply gtkmm2.json]: I missed these. Fixed.
